src/main/java/seedu/address/model/insurance/UniqueLifeInsuranceMap.java

```diff
@@ -0,0 +1,107 @@
+ */
+public class UniqueLifeInsuranceMap {
+
+    private final HashMap<UUID, LifeInsurance> internalMap = new HashMap<>();
+
+    /**
+     * Returns true if the map contains an equivalent UUID as the given argument.
+     */
+    public boolean containsKey(UUID toCheck) {
+        requireNonNull(toCheck);
+        return internalMap.containsKey(toCheck);
+    }
+
+    /**
+     * Returns true if the map contains an equivalent insurance as the given argument.
+     */
+    public boolean containsValue(ReadOnlyInsurance toCheck) {
+        requireNonNull(toCheck);
+        return internalMap.containsValue(toCheck);
+    }
+
+    /**
+     * Adds an life insurance to the map.
+     *
+     * @throws DuplicateInsuranceException if the life insurance to add is a duplicate of an
+     * existing life insurance in the list.
+     */
+    public void put(UUID key, ReadOnlyInsurance toPut) throws DuplicateInsuranceException {
+        requireNonNull(toPut);
+        if (containsValue(toPut)) {
+            throw new DuplicateInsuranceException();
+        }
+        internalMap.put(key, new LifeInsurance(toPut));
+    }
+
+    /**
+     * Retrieves an life insurance from the map.
+     *
+     * @throws InsuranceNotFoundException if the life insurance to add is a duplicate of an
+     * existing life insurance in the list.
+     */
+    public LifeInsurance get(UUID key) throws InsuranceNotFoundException {
+        requireNonNull(key);
+        if (!containsKey(key)) {
+            throw new InsuranceNotFoundException();
+        }
+        return internalMap.get(key);
+    }
+
+    public void forEach(BiConsumer<UUID, LifeInsurance> action) {
+        internalMap.forEach(action);
+    }
+
+    public void setInsurances(UniqueLifeInsuranceMap replacement) {
+        this.internalMap.clear();
+        this.internalMap.putAll(replacement.internalMap);
+    }
+
+    public void setInsurances(Map<UUID, ? extends ReadOnlyInsurance> insurances) throws DuplicateInsuranceException {
+        final UniqueLifeInsuranceMap replacement = new UniqueLifeInsuranceMap();
+        for (final Map.Entry<UUID, ? extends ReadOnlyInsurance> entry : insurances.entrySet()) {
+            replacement.put(entry.getKey(), entry.getValue());
+        }
+        setInsurances(replacement);
+    }
+
+    /**
+     * Returns the backing map as an unmodifiable {@code ObservableList}.
+     */
+    public Map<UUID, ReadOnlyInsurance> asMap() {
+        assert CollectionUtil.elementsAreUnique(internalMap.values());
+        return Collections.unmodifiableMap(internalMap);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof UniqueLifeInsuranceMap// instanceof handles nulls
+                && this.internalMap.equals(((UniqueLifeInsuranceMap) other).internalMap));
+    }
+
+    @Override
+    public int hashCode() {
+        return internalMap.hashCode();
+    }
+}
```

Comment:
wow this class is a really good solution well done!

Reply:
😄